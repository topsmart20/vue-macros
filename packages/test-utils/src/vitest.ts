import path from 'node:path'
import glob, { type Options as GlobOptions } from 'fast-glob'
import { describe, expect, test } from 'vitest'
import { normalizePath } from '@rollup/pluginutils'

interface Options {
  params?: [name: string, values?: any[]][]
  promise?: boolean
}

const SKIP_VUE2 = !!process.env.SKIP_VUE2

export async function testFixtures(
  globs: string | string[],
  exec: (args: Record<string, any>, id: string) => any,
  options?: GlobOptions & Options,
): Promise<void>
export async function testFixtures(
  files: Record<string, string>,
  exec: (args: Record<string, any>, id: string, code: string) => any,
  options?: Options,
): Promise<void>
export async function testFixtures(
  globsOrFiles: string | string[] | Record<string, string>,
  cb: (args: Record<string, any>, id: string, code: string) => any,
  { params, promise, ...globOptions }: GlobOptions & Options = {},
) {
  let files: Record<string, string | undefined>

  if (typeof globsOrFiles === 'string' || Array.isArray(globsOrFiles)) {
    files = Object.fromEntries(
      (await glob(globsOrFiles, globOptions)).map((file) => [file, undefined]),
    )
  } else {
    files = globsOrFiles
  }

  for (const [id, code] of Object.entries(files)) {
    makeTests(id, code, [[normalizePath(id)], ...(params || [])])()
  }

  function getName(name: string, value: any) {
    return value !== undefined ? `${name} = ${String(value)}` : name
  }

  function makeTests(
    id: string,
    code: string | undefined,
    params: NonNullable<Options['params']>,
    args: Record<string, any> = {},
  ) {
    const [currParams, ...restParams] = params
    const [name, values = [undefined]] = currParams
    if (restParams.length > 0) {
      return () => {
        for (const value of values) {
          const currArgs = { ...args, [name]: value }

          describe(
            getName(name, value),
            makeTests(id, code, restParams, currArgs),
          )
        }
      }
    } else {
      return () => {
        for (const value of values) {
          const testName = getName(name, value)
          const isSkip = testName.includes('vue2') && SKIP_VUE2

          test.skipIf(isSkip)(testName, async () => {
            const currArgs = { ...args, [name]: value }
            const execute = () =>
              cb(currArgs, path.resolve(globOptions.cwd || '/', id), code!)
            if (id.includes('error')) {
              if (promise) {
                await expect(execute()).rejects.toThrowErrorMatchingSnapshot()
              } else {
                expect(execute).toThrowErrorMatchingSnapshot()
              }
            } else if (promise) {
              await expect(
                (execute() as Promise<any>).catch((error) => {
                  console.warn(error)
                  return Promise.reject(error)
                }),
              ).resolves.toMatchSnapshot()
            } else {
              expect(execute()).toMatchSnapshot()
            }
          })
        }
      }
    }
  }
}
