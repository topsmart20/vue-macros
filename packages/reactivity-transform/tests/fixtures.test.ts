import { resolve } from 'node:path'
import { describe } from 'vitest'
import {
  RollupEsbuildPlugin,
  RollupVue,
  RollupVue2,
  rollupBuild,
  testFixtures,
} from '@vue-macros/test-utils'
import VueReactivityTransform from '../src/rollup'

describe('fixtures', async () => {
  await testFixtures(
    'tests/fixtures/**/*.{vue,js,ts}',
    (args, id) => {
      if (id.includes('vue2'))
        return rollupBuild(id, [
          VueReactivityTransform(),
          // @ts-expect-error rollup 4
          RollupVue2({
            compiler: require('vue2/compiler-sfc'),
          }),
          RollupEsbuildPlugin({
            target: 'esnext',
          }),
        ])
      else
        return rollupBuild(id, [
          VueReactivityTransform(),
          RollupVue({
            compiler: require('vue/compiler-sfc'),
          }),
          RollupEsbuildPlugin({
            target: 'esnext',
          }),
        ])
    },
    {
      cwd: resolve(__dirname, '..'),
      promise: true,
    }
  )
})
