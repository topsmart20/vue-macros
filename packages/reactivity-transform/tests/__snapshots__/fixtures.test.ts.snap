// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > tests/fixtures/vue2/basic.ts 1`] = `
"import { ref } from 'vue';

let foo = ref(\\"msg\\");
foo.value = \\"world\\";

export { foo };
"
`;

exports[`fixtures > tests/fixtures/vue2/basic.vue 1`] = `
"import { defineComponent, ref } from 'vue';
import __normalizer from '[NULL]plugin-vue2:normalizer';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"basic\\",
  setup(__props) {
    let foo = ref(\\"msg\\");
    foo.value = \\"world\\";
    return { __sfc: true, foo, _ref: ref };
  }
});

const _sfc_render = null;
const _sfc_staticRenderFns = null;
var __component__ = /* @__PURE__ */ __normalizer(
  _sfc_main,
  _sfc_render,
  _sfc_staticRenderFns,
  false,
  null,
  null,
  null,
  null
);
__FILE__;
var basic = __component__.exports;

export { basic as default };
"
`;

exports[`fixtures > tests/fixtures/vue2/destructure-props.vue 1`] = `
"import { defineComponent } from 'vue';
import { createPropsRestProxy } from '/vue-macros/reactivity-transform/helper';
import __normalizer from '[NULL]plugin-vue2:normalizer';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"destructure-props\\",
  props: {
    foo: { type: String, required: true, default: \\"foo\\" },
    bar: { type: Number, required: true, default: 10 },
    baz: { type: Boolean, required: true }
  },
  setup(__props) {
    const props = createPropsRestProxy(__props, [\\"foo\\", \\"bar\\"]);
    console.log(__props.foo, __props.bar, props);
    return { __sfc: true, props, createPropsRestProxy };
  }
});

const _sfc_render = null;
const _sfc_staticRenderFns = null;
var __component__ = /* @__PURE__ */ __normalizer(
  _sfc_main,
  _sfc_render,
  _sfc_staticRenderFns,
  false,
  null,
  null,
  null,
  null
);
__FILE__;
var destructureProps = __component__.exports;

export { destructureProps as default };
"
`;

exports[`fixtures > tests/fixtures/vue2/script-and-setup.vue 1`] = `
"import { ref, defineComponent } from 'vue';
import __normalizer from '[NULL]plugin-vue2:normalizer';

let foo = ref(\\"msg\\");
foo.value = \\"world\\";
var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"script-and-setup\\",
  setup(__props) {
    foo.value = \\"foo\\";
    return { __sfc: true, foo };
  }
});

var _sfc_render = function render() {
  var _vm = this, _c = _vm._self._c, _setup = _vm._self._setupProxy;
  return _c(\\"div\\", [_vm._v(\\" \\" + _vm._s(_setup.foo) + \\" \\")]);
};
var _sfc_staticRenderFns = [];
_sfc_render._withStripped = true;
var __component__ = /* @__PURE__ */ __normalizer(
  _sfc_main,
  _sfc_render,
  _sfc_staticRenderFns,
  false,
  null,
  null,
  null,
  null
);
__FILE__;
var scriptAndSetup = __component__.exports;

export { scriptAndSetup as default };
"
`;

exports[`fixtures > tests/fixtures/vue3/alias.vue 1`] = `
"import { toRef, toDisplayString } from 'vue';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

const _sfc_main = {
  __name: 'alias',
  props: {
  foo: String,
  baz: Number,
},
  setup(__props) {
const propAlias2 = toRef(__props, \\"baz\\");
const propAlias = toRef(__props, \\"foo\\");



return (_ctx, _cache) => {
  return toDisplayString(propAlias.value) + \\" \\" + toDisplayString(propAlias2.value)
}
}

};
var alias = /*#__PURE__*/_export_sfc(_sfc_main, [__FILE__]);

export { alias as default };
"
`;

exports[`fixtures > tests/fixtures/vue3/as.ts 1`] = `
"import { ref } from 'vue';

let foo = ref(\\"msg\\");
foo.value = \\"world\\";
console.log(foo);

export { foo };
"
`;

exports[`fixtures > tests/fixtures/vue3/basic.ts 1`] = `
"import { ref } from 'vue';

let foo = ref(\\"msg\\");
foo.value = \\"world\\";

export { foo };
"
`;

exports[`fixtures > tests/fixtures/vue3/basic.vue 1`] = `
"import { defineComponent, ref } from 'vue';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"basic\\",
  setup(__props) {
    let foo = ref(\\"msg\\");
    foo.value = \\"world\\";
    return () => {
    };
  }
});

var basic = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { basic as default };
"
`;

exports[`fixtures > tests/fixtures/vue3/default-value.vue 1`] = `
"import { defineComponent } from 'vue';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"default-value\\",
  props: {
    fn: { type: Function, required: true, default: () => {
      console.log(\\"hello\\");
    } },
    obj: { type: Object, required: true, default: () => ({ foo: \\"bar\\" }) },
    arr: { type: Array, required: true, default: () => [\\"foo\\", \\"bar\\"] }
  },
  setup(__props) {
    console.log(__props.fn, __props.obj, __props.arr);
    return () => {
    };
  }
});

var defaultValue = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { defaultValue as default };
"
`;

exports[`fixtures > tests/fixtures/vue3/destructure-props.vue 1`] = `
"import { defineComponent } from 'vue';
import { createPropsRestProxy } from '/vue-macros/reactivity-transform/helper';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"destructure-props\\",
  props: {
    foo: { type: String, required: true, default: \\"foo\\" },
    bar: { type: Number, required: true, default: 10 },
    baz: { type: Boolean, required: true }
  },
  setup(__props) {
    const props = createPropsRestProxy(__props, [\\"foo\\", \\"bar\\"]);
    console.log(__props.foo, __props.bar, props);
    return () => {
    };
  }
});

var destructureProps = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { destructureProps as default };
"
`;

exports[`fixtures > tests/fixtures/vue3/runtime-default.vue 1`] = `
"import { defineComponent, mergeDefaults } from 'vue';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"runtime-default\\",
  props: mergeDefaults([\\"foo\\"], { foo: \\"str\\" }),
  setup(__props) {
    console.log(__props.foo);
    return () => {
    };
  }
});

var runtimeDefault = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { runtimeDefault as default };
"
`;

exports[`fixtures > tests/fixtures/vue3/script-and-setup.vue 1`] = `
"import { ref, defineComponent, openBlock, createElementBlock, toDisplayString, unref } from 'vue';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

let foo = ref(\\"msg\\");
foo.value = \\"world\\";
var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"script-and-setup\\",
  setup(__props) {
    foo.value = \\"foo\\";
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(
        \\"div\\",
        null,
        toDisplayString(unref(foo)),
        1
        /* TEXT */
      );
    };
  }
});

var scriptAndSetup = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { scriptAndSetup as default };
"
`;
